File: 351004/Sapeshko/lab1-6/src/main/java/com/example/demo/security/JwtService.java
package com.example.demo.security;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.context.annotation.Profile;
import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.util.Date;

@Service
@Profile("docker")
public class JwtService {

    private final SecretKey signingKey;
    private final long expirationMs;

    public JwtService(@Value("${jwt.secret:mysupersecretkeywhichisatleast256bitslong!}") String secret,
                      @Value("${jwt.expiration:3600000}") long expirationMs) {
        this.signingKey = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
        this.expirationMs = expirationMs;
    }

    public String generateToken(String login, String role) {
        return Jwts.builder()
                .subject(login)
                .claim("role", role)
                .issuedAt(new Date())
                .expiration(new Date(System.currentTimeMillis() + expirationMs))
                .signWith(signingKey)
                .compact();
    }

    public Claims validateToken(String token) {
        return Jwts.parser()
                .verifyWith(signingKey)
                .build()
                .parseSignedClaims(token)
                .getPayload();
    }
}

File: 351002/Abrashin/publisher/src/main/kotlin/com/example/distcomp/security/JwtService.kt
package com.example.distcomp.security

import com.example.distcomp.model.Creator
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service
import java.util.Date
import javax.crypto.SecretKey

@Service
class JwtService(
    @Value("\${security.jwt.secret}") secret: String,
    @Value("\${security.jwt.expiration-ms}") private val expirationMs: Long
) {
    private val signingKey: SecretKey = Keys.hmacShaKeyFor(resolveKey(secret))

    fun generateToken(creator: Creator): String {
        val now = Date()
        val expiresAt = Date(now.time + expirationMs)
        return Jwts.builder()
            .subject(creator.login ?: throw IllegalArgumentException("Creator login is required for JWT"))
            .issuedAt(now)
            .expiration(expiresAt)
            .claim("role", creator.role.name)
            .signWith(signingKey)
            .compact()
    }

    fun extractLogin(token: String): String = extractAllClaims(token).subject

    fun isTokenValid(token: String, userDetails: UserDetails): Boolean {
        val claims = extractAllClaims(token)
        val tokenRole = claims["role"]?.toString()
        val currentRole = userDetails.authorities.firstOrNull()?.authority?.removePrefix("ROLE_")
        return claims.subject == userDetails.username &&
            !claims.expiration.before(Date()) &&
            tokenRole == currentRole
    }

    private fun extractAllClaims(token: String): Claims =
        Jwts.parser()
            .verifyWith(signingKey)
            .build()
            .parseSignedClaims(token)
            .payload

    private fun resolveKey(secret: String): ByteArray =
        runCatching { Decoders.BASE64.decode(secret) }
            .getOrElse { secret.toByteArray(Charsets.UTF_8) }
}
